libs/client-api/src/collab_sync/collab_sink.rs

```diff
@@ -357,19 +366,34 @@ where
 
         // Try to merge the next message with the last message. Only merge when:
         // 1. The last message is not in the flying messages.
-        // 2. The last message can be merged.
+        // 2. The last message can be merged and the next message can be merged.
         // 3. The last message's payload size is less than the maximum payload size.
         if let Some(last) = items.last_mut() {
-          if !sending_messages.contains(&last.msg_id())
+          let can_merge = !sending_messages.contains(&last.msg_id())
             && last.message().payload_size() < self.config.maximum_payload_size
             && last.mergeable()
-            && last.merge(&next, &self.config.maximum_payload_size).is_ok()
-          {
+            && next.mergeable()
```

Comment:
When merging the last and next, **we should check if the next is mergeable**. If the last is mergeable but the next is NOT, they should not be merged.